<a href="https://colab.research.google.com/github/virajsoni8899/virajsoni8899/blob/main/Diamond_price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Attributes
- price: price in us dollars
- carat: weight of the diamond
- cut: cut quality
- color: color of the diamond
- clarity: clarity of the diamond
- X: lenght in mm
- Y: width in mm
- Z: depth in mm
- depth: total dpeth in percentage 100*z/mean(x,y)
- table: width of the top of the diamond relative to the widest point

In [1]:
#step 1 import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#step 2 read dataset
diamond = pd.read_csv("https://raw.githubusercontent.com/YBIFoundation/Dataset/main/Diamonds.csv")
diamond.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [3]:
diamond.describe()

carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

In [4]:
diamond.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    53940 non-null  float64
 1   cut      53940 non-null  object 
 2   color    53940 non-null  object 
 3   clarity  53940 non-null  object 
 4   depth    53940 non-null  float64
 5   table    53940 non-null  float64
 6   price    53940 non-null  int64  
 7   x        53940 non-null  float64
 8   y        53940 non-null  float64
 9   z        53940 non-null  float64
dtypes: float64(6), int64(1), object(3)
memory usage: 4.1+ MB


In [5]:
#encoding
diamond.cut.value_counts()


Ideal        21551
Premium      13791
Very Good    12082
Good          4906
Fair          1610
Name: cut, dtype: int64

In [6]:
diamond.cut.replace({'Ideal':0,'Fair':1,'Good':2,'Very Good':3,'Premium':4},inplace=True)

In [7]:
diamond.cut.value_counts()

0    21551
4    13791
3    12082
2     4906
1     1610
Name: cut, dtype: int64

In [8]:
diamond.color.value_counts()

G    11292
E     9797
F     9542
H     8304
D     6775
I     5422
J     2808
Name: color, dtype: int64

In [9]:
diamond.color.replace({'D':6,'E':5,'F':4,'G':3,'H':2,'I':1,'J':0},inplace=True)

In [10]:
diamond.color.value_counts()

3    11292
5     9797
4     9542
2     8304
6     6775
1     5422
0     2808
Name: color, dtype: int64

- Flawless (FL) No inclusions and no blemishes visible under 10x magnification
- Internally Flawless (IF) No inclusions visible under 10x magnification
- Very, Very Slightly Included (VVS1 and VVS2) Inclusions so slight they are difficult for a skilled grader to see under 10x magnification
- Very Slightly Included (VS1 and VS2) Inclusions are observed with effort under 10x magnification, but can be characterized as minor
- Slightly Included (SI1 and SI2) Inclusions are noticeable under 10x magnification
- Included (I1, I2, and I3) Inclusions are obvious under 10x magnification which may affect transparency and brilliance

In [11]:
diamond.clarity.value_counts()

SI1     13065
VS2     12258
SI2      9194
VS1      8171
VVS2     5066
VVS1     3655
IF       1790
I1        741
Name: clarity, dtype: int64

In [12]:
diamond.clarity.replace({'IF':7,'VVS1':6,'VVS2':5,'VS1':4,'VS2':3,'SI1':2,'SI2':1,'I1':0},inplace=True)

In [13]:
diamond.clarity.value_counts()

2    13065
3    12258
1     9194
4     8171
5     5066
6     3655
7     1790
0      741
Name: clarity, dtype: int64

In [14]:
diamond.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    53940 non-null  float64
 1   cut      53940 non-null  int64  
 2   color    53940 non-null  int64  
 3   clarity  53940 non-null  int64  
 4   depth    53940 non-null  float64
 5   table    53940 non-null  float64
 6   price    53940 non-null  int64  
 7   x        53940 non-null  float64
 8   y        53940 non-null  float64
 9   z        53940 non-null  float64
dtypes: float64(6), int64(4)
memory usage: 4.1 MB


In [15]:
#defining x & y
diamond.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'price', 'x', 'y',
       'z'],
      dtype='object')

In [16]:
y = diamond['price']
x = diamond[['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'x', 'y',
       'z']]

In [17]:
#split data
from sklearn.model_selection import train_test_split
x_train,y_train,x_test,y_test = train_test_split(x,y,random_state=2529)

In [18]:
#select a model
import statsmodels.api as sm


In [19]:
x1 = sm.add_constant(x)


In [20]:
model = sm.OLS(y,x)

In [21]:
print(model.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                  price   R-squared (uncentered):                   0.952
Model:                            OLS   Adj. R-squared (uncentered):              0.952
Method:                 Least Squares   F-statistic:                          1.190e+05
Date:                Wed, 13 Mar 2024   Prob (F-statistic):                        0.00
Time:                        14:54:20   Log-Likelihood:                     -4.6017e+05
No. Observations:               53940   AIC:                                  9.204e+05
Df Residuals:                   53931   BIC:                                  9.204e+05
Df Model:                           9                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------


This output shows the results of a regression analysis used to predict the price of diamonds based on several characteristics. The model indicates that about 95.2% of the variation in diamond prices can be explained by the variables included. The coefficients for each variable represent the expected change in price for a one-unit change in that variable. The p-values associated with each coefficient indicate whether each variable is statistically significant in predicting price. Overall, the model appears to be statistically significant, but some variables may not be useful predictors.





